### PR TITLE
Updated Micro epsilon SDK and dependencies

### LINKED ISSUE #1: failing to install scanCONTROL
when trying to install micro epsilon scancontrol it creates the following messages:

Install Micro Epsilon scanCONTROL API (y/n)?y
Architecture: x86_64
ln: failed to create symbolic link '/usr/local/lib/libllt.so': File exists
ln: failed to create symbolic link '/usr/local/lib/libmescan.so': File exists

I tried to catkin_make the workspace anyways, but that resulted in the following error messages:

/home/joerivos/Documents/catkin_ws/src/micro_epsilon_scancontrol-develop/scancontrol/src/driver.cpp: In constructor ‘scancontrol_driver::ScanControlDriver::ScanControlDriver(ros::NodeHandle, ros::NodeHandle)’:
/home/joerivos/Documents/catkin_ws/src/micro_epsilon_scancontrol-develop/scancontrol/src/driver.cpp:185:13: error: ‘return_value’ was not declared in this scope
         if (return_value = SetPartialProfile(&config_.resolution) < GENERAL_FUNCTION_OK){
             ^~~~~~~~~~~~
/home/joerivos/Documents/catkin_ws/src/micro_epsilon_scancontrol-develop/scancontrol/src/driver.cpp:185:13: note: suggested alternative: ‘return_code’
         if (return_value = SetPartialProfile(&config_.resolution) < GENERAL_FUNCTION_OK){
             ^~~~~~~~~~~~
             return_code
/home/joerivos/Documents/catkin_ws/src/micro_epsilon_scancontrol-develop/scancontrol/src/driver.cpp:185:46: error: invalid conversion from ‘int*’ to ‘int’ [-fpermissive]
         if (return_value = SetPartialProfile(&config_.resolution) < GENERAL_FUNCTION_OK){
                                              ^~~~~~~~~~~~~~~~~~~
In file included from /home/joerivos/Documents/catkin_ws/src/micro_epsilon_scancontrol-develop/scancontrol/src/driver.cpp:1:0:
/home/joerivos/Documents/catkin_ws/src/micro_epsilon_scancontrol-develop/scancontrol/include/scancontrol/driver.h:34:17: note:   initializing argument 1 of ‘int scancontrol_driver::ScanControlDriver::SetPartialProfile(int&)’
             int SetPartialProfile(int &resolution);
                 ^~~~~~~~~~~~~~~~~
/home/joerivos/Documents/catkin_ws/src/micro_epsilon_scancontrol-develop/scancontrol/src/driver.cpp:185:46: error: cannot bind rvalue ‘(int)(&((scancontrol_driver::ScanControlDriver*)this)->scancontrol_driver::ScanControlDriver::config_.scancontrol_driver::ScanControlDriver::<unnamed struct>::resolution)’ to ‘int&’
         if (return_value = SetPartialProfile(&config_.resolution) < GENERAL_FUNCTION_OK){
                                              ^~~~~~~~~~~~~~~~~~~
/home/joerivos/Documents/catkin_ws/src/micro_epsilon_scancontrol-develop/scancontrol/src/driver.cpp: In member function ‘int scancontrol_driver::ScanControlDriver::SetPartialProfile(int&)’:
/home/joerivos/Documents/catkin_ws/src/micro_epsilon_scancontrol-develop/scancontrol/src/driver.cpp:221:14: error: ‘return_value’ was not declared in this scope
         if ((return_value = device_interface.SetProfileConfig(PARTIAL_PROFILE)) < GENERAL_FUNCTION_OK){
              ^~~~~~~~~~~~
/home/joerivos/Documents/catkin_ws/src/micro_epsilon_scancontrol-develop/scancontrol/src/driver.cpp:221:14: note: suggested alternative: ‘si_value’
         if ((return_value = device_interface.SetProfileConfig(PARTIAL_PROFILE)) < GENERAL_FUNCTION_OK){
              ^~~~~~~~~~~~
              si_value
/home/joerivos/Documents/catkin_ws/src/micro_epsilon_scancontrol-develop/scancontrol/src/driver.cpp:233:14: error: ‘return_value’ was not declared in this scope
         if ((return_value = device_interface.SetPartialProfile(&t_partial_profile_)) < GENERAL_FUNCTION_OK){
              ^~~~~~~~~~~~
/home/joerivos/Documents/catkin_ws/src/micro_epsilon_scancontrol-develop/scancontrol/src/driver.cpp:233:14: note: suggested alternative: ‘si_value’
         if ((return_value = device_interface.SetPartialProfile(&t_partial_profile_)) < GENERAL_FUNCTION_OK){
              ^~~~~~~~~~~~
              si_value
/home/joerivos/Documents/catkin_ws/src/micro_epsilon_scancontrol-develop/scancontrol/src/driver.cpp: In member function ‘bool scancontrol_driver::ScanControlDriver::ServiceSetResolution(scancontrol::SetResolution::Request&, scancontrol::SetResolution::Response&)’:
/home/joerivos/Documents/catkin_ws/src/micro_epsilon_scancontrol-develop/scancontrol/src/driver.cpp:380:54: error: invalid conversion from ‘scancontrol::SetResolutionRequest_<std::allocator<void> >::_resolution_type* {aka unsigned int*}’ to ‘int’ [-fpermissive]
         if (response.return_code = SetPartialProfile(&(request.resolution)) < GENERAL_FUNCTION_OK){
                                                      ^~~~~~~~~~~~~~~~~~~~~
/home/joerivos/Documents/catkin_ws/src/micro_epsilon_scancontrol-develop/scancontrol/src/driver.cpp:219:9: note:   initializing argument 1 of ‘int scancontrol_driver::ScanControlDriver::SetPartialProfile(int&)’
     int ScanControlDriver::SetPartialProfile(int &resolution){
         ^~~~~~~~~~~~~~~~~
/home/joerivos/Documents/catkin_ws/src/micro_epsilon_scancontrol-develop/scancontrol/src/driver.cpp:380:54: error: cannot bind rvalue ‘(int)(& request.scancontrol::SetResolutionRequest_<std::allocator<void> >::resolution)’ to ‘int&’
         if (response.return_code = SetPartialProfile(&(request.resolution)) < GENERAL_FUNCTION_OK){
                                                      ^~~~~~~~~~~~~~~~~~~~~
micro_epsilon_scancontrol-develop/scancontrol/CMakeFiles/driver_nodelet.dir/build.make:86: recipe for target 'micro_epsilon_scancontrol-develop/scancontrol/CMakeFiles/driver_nodelet.dir/src/driver.cpp.o' failed
make[2]: *** [micro_epsilon_scancontrol-develop/scancontrol/CMakeFiles/driver_nodelet.dir/src/driver.cpp.o] Error 1
make[2]: *** Waiting for unfinished jobs....
Scanning dependencies of target scancontrol_generate_messages
[ 97%] Built target scancontrol_generate_messages
CMakeFiles/Makefile2:1587: recipe for target 'micro_epsilon_scancontrol-develop/scancontrol/CMakeFiles/driver_nodelet.dir/all' failed
make[1]: *** [micro_epsilon_scancontrol-develop/scancontrol/CMakeFiles/driver_nodelet.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
/home/joerivos/Documents/catkin_ws/src/micro_epsilon_scancontrol-develop/scancontrol/src/driver.cpp: In constructor ‘scancontrol_driver::ScanControlDriver::ScanControlDriver(ros::NodeHandle, ros::NodeHandle)’:
/home/joerivos/Documents/catkin_ws/src/micro_epsilon_scancontrol-develop/scancontrol/src/driver.cpp:185:13: error: ‘return_value’ was not declared in this scope
         if (return_value = SetPartialProfile(&config_.resolution) < GENERAL_FUNCTION_OK){
             ^~~~~~~~~~~~
/home/joerivos/Documents/catkin_ws/src/micro_epsilon_scancontrol-develop/scancontrol/src/driver.cpp:185:13: note: suggested alternative: ‘return_code’
         if (return_value = SetPartialProfile(&config_.resolution) < GENERAL_FUNCTION_OK){
             ^~~~~~~~~~~~
             return_code
/home/joerivos/Documents/catkin_ws/src/micro_epsilon_scancontrol-develop/scancontrol/src/driver.cpp:185:46: error: invalid conversion from ‘int*’ to ‘int’ [-fpermissive]
         if (return_value = SetPartialProfile(&config_.resolution) < GENERAL_FUNCTION_OK){
                                              ^~~~~~~~~~~~~~~~~~~
In file included from /home/joerivos/Documents/catkin_ws/src/micro_epsilon_scancontrol-develop/scancontrol/src/driver.cpp:1:0:
/home/joerivos/Documents/catkin_ws/src/micro_epsilon_scancontrol-develop/scancontrol/include/scancontrol/driver.h:34:17: note:   initializing argument 1 of ‘int scancontrol_driver::ScanControlDriver::SetPartialProfile(int&)’
             int SetPartialProfile(int &resolution);
                 ^~~~~~~~~~~~~~~~~
/home/joerivos/Documents/catkin_ws/src/micro_epsilon_scancontrol-develop/scancontrol/src/driver.cpp:185:46: error: cannot bind rvalue ‘(int)(&((scancontrol_driver::ScanControlDriver*)this)->scancontrol_driver::ScanControlDriver::config_.scancontrol_driver::ScanControlDriver::<unnamed struct>::resolution)’ to ‘int&’
         if (return_value = SetPartialProfile(&config_.resolution) < GENERAL_FUNCTION_OK){
                                              ^~~~~~~~~~~~~~~~~~~
/home/joerivos/Documents/catkin_ws/src/micro_epsilon_scancontrol-develop/scancontrol/src/driver.cpp: In member function ‘int scancontrol_driver::ScanControlDriver::SetPartialProfile(int&)’:
/home/joerivos/Documents/catkin_ws/src/micro_epsilon_scancontrol-develop/scancontrol/src/driver.cpp:221:14: error: ‘return_value’ was not declared in this scope
         if ((return_value = device_interface.SetProfileConfig(PARTIAL_PROFILE)) < GENERAL_FUNCTION_OK){
              ^~~~~~~~~~~~
/home/joerivos/Documents/catkin_ws/src/micro_epsilon_scancontrol-develop/scancontrol/src/driver.cpp:221:14: note: suggested alternative: ‘si_value’
         if ((return_value = device_interface.SetProfileConfig(PARTIAL_PROFILE)) < GENERAL_FUNCTION_OK){
              ^~~~~~~~~~~~
              si_value
/home/joerivos/Documents/catkin_ws/src/micro_epsilon_scancontrol-develop/scancontrol/src/driver.cpp:233:14: error: ‘return_value’ was not declared in this scope
         if ((return_value = device_interface.SetPartialProfile(&t_partial_profile_)) < GENERAL_FUNCTION_OK){
              ^~~~~~~~~~~~
/home/joerivos/Documents/catkin_ws/src/micro_epsilon_scancontrol-develop/scancontrol/src/driver.cpp:233:14: note: suggested alternative: ‘si_value’
         if ((return_value = device_interface.SetPartialProfile(&t_partial_profile_)) < GENERAL_FUNCTION_OK){
              ^~~~~~~~~~~~
              si_value
/home/joerivos/Documents/catkin_ws/src/micro_epsilon_scancontrol-develop/scancontrol/src/driver.cpp: In member function ‘bool scancontrol_driver::ScanControlDriver::ServiceSetResolution(scancontrol::SetResolution::Request&, scancontrol::SetResolution::Response&)’:
/home/joerivos/Documents/catkin_ws/src/micro_epsilon_scancontrol-develop/scancontrol/src/driver.cpp:380:54: error: invalid conversion from ‘scancontrol::SetResolutionRequest_<std::allocator<void> >::_resolution_type* {aka unsigned int*}’ to ‘int’ [-fpermissive]
         if (response.return_code = SetPartialProfile(&(request.resolution)) < GENERAL_FUNCTION_OK){
                                                      ^~~~~~~~~~~~~~~~~~~~~
/home/joerivos/Documents/catkin_ws/src/micro_epsilon_scancontrol-develop/scancontrol/src/driver.cpp:219:9: note:   initializing argument 1 of ‘int scancontrol_driver::ScanControlDriver::SetPartialProfile(int&)’
     int ScanControlDriver::SetPartialProfile(int &resolution){
         ^~~~~~~~~~~~~~~~~
/home/joerivos/Documents/catkin_ws/src/micro_epsilon_scancontrol-develop/scancontrol/src/driver.cpp:380:54: error: cannot bind rvalue ‘(int)(& request.scancontrol::SetResolutionRequest_<std::allocator<void> >::resolution)’ to ‘int&’
         if (response.return_code = SetPartialProfile(&(request.resolution)) < GENERAL_FUNCTION_OK){
                                                      ^~~~~~~~~~~~~~~~~~~~~
micro_epsilon_scancontrol-develop/scancontrol/CMakeFiles/driver_node.dir/build.make:86: recipe for target 'micro_epsilon_scancontrol-develop/scancontrol/CMakeFiles/driver_node.dir/src/driver.cpp.o' failed
make[2]: *** [micro_epsilon_scancontrol-develop/scancontrol/CMakeFiles/driver_node.dir/src/driver.cpp.o] Error 1
CMakeFiles/Makefile2:1624: recipe for target 'micro_epsilon_scancontrol-develop/scancontrol/CMakeFiles/driver_node.dir/all' failed
make[1]: *** [micro_epsilon_scancontrol-develop/scancontrol/CMakeFiles/driver_node.dir/all] Error 2
Makefile:140: recipe for target 'all' failed
make: *** [all] Error 2
Invoking "make -j4 -l4" failed

### PR DESCRIPTION
**Pull request by NithishkumarS**
Update the driver and Dockerfile used for build testing to v1.0.0 of the SDK (currently the latest version) for the Micro Epsilon scanCONTROL devices.

Jira issue: [LH2-210](https://samxl.atlassian.net/browse/LH2-210)

## Motivation and Context

The current SDK version (v0.2.5) is outdated and does not support the latest devices. 

## Changes
- Update the Dockerfile to use the latest SDK version
  - Updated scanCONTROL SDK to v1.0.0 
  - Updated Aravis to 0.8.30

## Type of changes

- Breaking change (Updated the docker image to use Ubuntu focal and ros noetic)

## Checklist

- [x] Update version
- [x] Update dependencies
- [x] Test Docker build

### CODE REVIEW

File: .docker/Dockerfile [hunk 1]
```diff
@@ -13,45 +13,50 @@ RUN apt-get update && \
...
 
-RUN wget http://ftp.acc.umu.se/pub/GNOME/sources/aravis/0.6/aravis-0.6.4.tar.xz && \
-    tar xfJ aravis-0.6.4.tar.xz && \
-    rm aravis-0.6.4.tar.xz && \
-    cd aravis-0.6.4 && \
-    ./configure && \
-    make && \
-    checkinstall --pkgname aravis --requires="libglib2.0-dev, libxml2-dev" && \
-    ldconfig
+RUN pip3 install meson
```
Comment (dave992) [inline .docker/Dockerfile:27]: Run this as part of the run command starting on line 8. as that one installs all dependencies already. Also use the `--no-cache` option to reduce the image size slightly.

File: .docker/Dockerfile [hunk 2]
```diff
@@ -13,45 +13,50 @@ RUN apt-get update && \
...
+
+RUN wget https://github.com/AravisProject/aravis/releases/download/0.8.30/aravis-0.8.30.tar.xz -O aravis-0.8.30.tar.xz &&\
+    tar xfJ aravis-0.8.30.tar.xz && \
+    rm aravis-0.8.30.tar.xz && \
+    cd aravis-0.8.30 && \
+    meson setup build && \
+    cd build && \
+    ninja && \
+    checkinstall --pkgname aravis --pkgversion 0.8.30 ninja install
```
Comment (dave992) [inline .docker/Dockerfile:36]: Why did you remove the `--requires` tag? That makes sure the package manager installs those if not installed on your system.

File: .docker/Dockerfile [hunk 3]
```diff
@@ -13,45 +13,50 @@ RUN apt-get update && \
...
-    unzip scanCONTROL-Linux-SDK-${SCANCONTROL_SDK_VERSION//./-}.zip && \
+RUN wget https://software.micro-epsilon.com/scanCONTROL-Linux-SDK-${SCANCONTROL_SDK_VERSION//./-}.zip -O scanCONTROL-Linux-SDK-${SCANCONTROL_SDK_VERSION//./-}.zip && \
+    unzip scanCONTROL-Linux-SDK-${SCANCONTROL_SDK_VERSION//./-}.zip -d scanCONTROL-Linux-SDK/ && \
     rm scanCONTROL-Linux-SDK-${SCANCONTROL_SDK_VERSION//./-}.zip && \
-    cd /scanCONTROL\ Linux\ SDK\ ${SCANCONTROL_SDK_VERSION}/libmescan/ && \
+    cd scanCONTROL-Linux-SDK/libmescan/ && \ 
     meson builddir && \
     cd builddir && \
     ninja && \
     checkinstall --pkgname mescan --pkgversion ${SCANCONTROL_SDK_VERSION} --requires="aravis \(\>= 0.6.0\)" ninja install && \
```
Comment (dave992) [inline .docker/Dockerfile:46]: Does the aravis min version need an update?

File: .docker/Dockerfile [hunk 4]
```diff
@@ -13,45 +13,50 @@ RUN apt-get update && \
...
     ldconfig && \
-    cd "/scanCONTROL Linux SDK ${SCANCONTROL_SDK_VERSION}/libllt/" && \
+    cd ../../libllt && \
```
Comment (dave992) [inline .docker/Dockerfile:48]: Would keep this as an absolute path

**Review by dave992 — CHANGES_REQUESTED** [review]

File: .docker/Dockerfile (hunk 2 above)
Comment (NithishkumarS) [reply .docker/Dockerfile:36]: This originally was using the configure install. But aravis repo has updated the build instruction to use ninja build. Will add the

File: .docker/Dockerfile [hunk 5]
```diff
@@ -59,9 +67,10 @@ RUN mkdir /library_pkgs && \
 FROM ros:noetic-ros-core
 ARG SCANCONTROL_SDK_VERSION
 
-RUN apt-get update && \
-    apt-get install -y \
+RUN apt-get update && apt-get install -y --no-install-recommends\
       intltool \
+      libglib2.0-dev \
+      libusb-1.0-0-dev \
```
Comment (dave992) [inline .docker/Dockerfile:73]: This should not be needed if the `--required` tag is used correctly when creating the `.deb` file.

File: .docker/Dockerfile (hunk 1 above)
Comment (NithishkumarS) [reply .docker/Dockerfile]: Fixed

File: .docker/Dockerfile (hunk 3 above)
Comment (NithishkumarS) [reply .docker/Dockerfile]: Fixed

File: .docker/Dockerfile (hunk 4 above)
Comment (NithishkumarS) [reply .docker/Dockerfile]: Fixed

**Review by dave992 — APPROVED** [review]

File: .docker/Dockerfile [hunk 6]
```diff
@@ -1,68 +1,82 @@
...
-    checkinstall --pkgname aravis --requires="libglib2.0-dev, libxml2-dev" && \
-    ldconfig
+RUN wget https://github.com/AravisProject/aravis/releases/download/0.8.30/aravis-0.8.30.tar.xz -O aravis-0.8.30.tar.xz &&\
+    tar xfJ aravis-0.8.30.tar.xz && \
+    rm aravis-0.8.30.tar.xz && \
+    cd aravis-0.8.30 && \
+    meson setup build && \
+    cd build && \
+    ninja && \
+    checkinstall --pkgname aravis --pkgversion 0.8.30 --requires="libglib2.0-dev, libxml2-dev" ninja install
```
Comment (gavanderhoorn) [inline .docker/Dockerfile:44]: just curious why we 'bother' with `checkinstall` if this is a `Dockerfile`?

File: .docker/Dockerfile [hunk 7]
```diff
@@ -1,68 +1,82 @@
...
       libxml2-dev \
-      meson \
+      python3-pip \
+      ninja-build \
       pkg-config \
       unzip \
       wget \
-    && rm -rf /var/lib/apt/lists/*
+      xsltproc \
+    && rm -rf /var/lib/apt/lists/* && \
```
Comment (katettudelft) [inline .docker/Dockerfile:34]: I'd suggest expanding this a bit further - this is the standard 'clean-up' stanza I use:

```
apt-get autoremove -y &&\
apt-get clean -y &&\
rm -rf /var/lib/apt/lists/*
```

File: .docker/Dockerfile [hunk 8]
```diff
@@ -1,68 +1,82 @@
...
+RUN apt-get update && apt-get install -y --no-install-recommends\
       intltool \
       pkg-config \
     && rm -rf /var/lib/apt/lists/*
 
 COPY --from=build ["/library_pkgs", "/library_pkgs"]
 
 RUN apt-get update && \
-    apt install -y /library_pkgs/aravis_0.6.4-1_amd64.deb && \
+    apt install -y /library_pkgs/aravis_0.8.30-1_amd64.deb && \
```
Comment (katettudelft) [inline .docker/Dockerfile:78]: It's preferable to use `apt-get` rather than `apt` in scripts - it has a more reliable command line interface.

I think you can do all of the packages in one install command too (but please test it works if you change it).

Finally, if these are the only files in `/library_pkgs/`, then you could consider using
`apt-get install /library_pkgs/*` to make the command shorter. Obviously if there are additional files in there that you don't want, or the contents are subject to change, then don't do this. :)
